### PR TITLE
Tm2 upgrade ncne withdrawl

### PR DESCRIPTION
**Pull request by SamirHassoun**
- Updated the Withdrawal workflow with new stages
- Code refactoring

### CODE REVIEW

File: src/NCNEPortal/Helpers/StageTypeFactory.cs [hunk 1]
```diff
@@ -13,32 +13,92 @@ public StageTypeFactory(NcneWorkflowDbContext dbContext)
...
-                        .Where(t => t.TaskStageTypeId > (int)NcneTaskStageType.With_Geodesy
-                                             && t.TaskStageTypeId < (int)NcneTaskStageType.Withdrawal_action).ToList();
+                    stageIds = GetStagesForNewTaskNonAdoption();
                 }
             }
 
+            var taskStageTypes = _dbContext.TaskStageType
+                .Where(t => stageIds.Contains(t.TaskStageTypeId)).ToList();
+
+
```
Comment (umarkumark) [inline src/NCNEPortal/Helpers/StageTypeFactory.cs:38]: Remove extra line break

File: src/NCNEPortal/Helpers/StageTypeFactory.cs [hunk 2]
```diff
@@ -13,32 +13,92 @@ public StageTypeFactory(NcneWorkflowDbContext dbContext)
...
+                (int)NcneTaskStageType.Commit_To_Print,
+                (int)NcneTaskStageType.CIS,
+                (int)NcneTaskStageType.Publication,
+                (int)NcneTaskStageType.Publish_Chart,
+                (int)NcneTaskStageType.Clear_Vector,
+                (int)NcneTaskStageType.Retire_Old_Version,
+                (int)NcneTaskStageType.Consider_Withdrawn_Charts
+            };
+        }
+
```
Comment (umarkumark) [inline src/NCNEPortal/Helpers/StageTypeFactory.cs:105]: Remove line break

File: src/NCNEPortal/Helpers/WorkflowStageHelper.cs [hunk 3]
```diff
@@ -74,6 +97,7 @@ public List<NcneTaskStageType> GetNextStagesForCompletion(NcneTaskStageType curr
                     break;
                 case NcneTaskStageType.Consider_Withdrawn_Charts:
                     break;
+
```
Comment (umarkumark) [inline src/NCNEPortal/Helpers/WorkflowStageHelper.cs:100]: Remove line break

File: src/NCNEPortal/Pages/Workflow.cshtml.cs [hunk 4]
```diff
@@ -1066,6 +1066,16 @@ private void UpdateRoles(TaskInfo task, TaskRole role)
...
+                    NcneTaskStageType.Withdraw_Impact_Assessment => role.Compiler,
+                    NcneTaskStageType.Initial_Announcement => role.Compiler,
+                    NcneTaskStageType.Verify_Withdraw_Impact_Assessment => role.VerifierOne,
+                    NcneTaskStageType.Draft_Miscellaneous_Action_Email => role.Compiler,
+                    NcneTaskStageType.Verify_and_Complete_Miscellaneous_Action => role.VerifierOne,
+                    NcneTaskStageType.Consider_Consequential_Action => role.Compiler,
+                    NcneTaskStageType.Confirmation_Announcement => role.VerifierOne,
+                    NcneTaskStageType.CIS => role.Compiler,
+                    NcneTaskStageType.PMC_withdrawal => role.VerifierOne,
+
```
Comment (umarkumark) [inline src/NCNEPortal/Pages/Workflow.cshtml.cs]: Remove line break

File: src/NCNEPortal.UnitTests/PageValidationTests.cs [hunk 5]
```diff
@@ -622,7 +622,7 @@ public void Validation_for_ValidateForCompleteWorkflow_with_Invalid_user_fails(s
             var validationErrorMessages = new List<string>();
 
             var result =
-                _pageValidationHelper.ValidateForCompleteWorkflow(assignedUser, currentUser, validationErrorMessages);
+                _pageValidationHelper.ValidateForCompleteWorkflow(assignedUser, currentUser,false, validationErrorMessages);
```
Comment (umarkumark) [inline src/NCNEPortal.UnitTests/PageValidationTests.cs:625]: Space before false

File: src/NCNEPortal.UnitTests/IndexPageTests.cs [hunk 6]
```diff
@@ -25,6 +26,8 @@ public class IndexPageTests
         private List<TaskStage> _taskStages;
         private ICarisProjectHelper _fakecarisProjectHelper;
         private IOptions<GeneralConfig> _fakegeneralConfig;
+        private IWorkflowStageHelper _fakeWorkflowStageHelper;
+
```
Comment (umarkumark) [inline src/NCNEPortal.UnitTests/IndexPageTests.cs:30]: Remove line break

File: src/NCNEPortal/Helpers/StageTypeFactory.cs (hunk 1 above)
Comment (SamirHassoun) [reply src/NCNEPortal/Helpers/StageTypeFactory.cs]: Fixed

File: src/NCNEPortal/Helpers/WorkflowStageHelper.cs (hunk 3 above)
Comment (SamirHassoun) [reply src/NCNEPortal/Helpers/WorkflowStageHelper.cs]: Fixed

File: src/NCNEPortal.UnitTests/IndexPageTests.cs (hunk 6 above)
Comment (SamirHassoun) [reply src/NCNEPortal.UnitTests/IndexPageTests.cs:30]: Fixed

File: src/NCNEPortal.UnitTests/PageValidationTests.cs (hunk 5 above)
Comment (SamirHassoun) [reply src/NCNEPortal.UnitTests/PageValidationTests.cs]: fixed

**Review by umarkumark — APPROVED** [review]

File: src/NCNEPortal/Data/TaskStageType.json [hunk 7]
```diff
@@ -116,13 +116,55 @@
...
+    "SequenceNumber": 15,
+    "AllowRework": false
+  },
+  {
+    "TaskStageTypeId": 24,
+    "Name": "Verify Withdraw Impact Assessment",
+    "SequenceNumber": 16,
+    "AllowRework": false
+  },
+  {
```
Comment (uhakdt) [inline src/NCNEPortal/Data/TaskStageType.json:146]: Why is this repeated?
In the PBI it says: "Verify and complete Miscellaneous Action"
So should that not be Verify Miscellaneous Action Email and Complete Miscellaneous Action Email or just 1 or maybe the draft is also needed. Not sure.

Comment (uhakdt) [reply src/NCNEPortal/Data/TaskStageType.json:146]: Actually as I see below now - the second, repeated one should probably be "Verify and complete Miscellaneous Action"

File: src/NCNEPortal/Helpers/PageValidationHelper.cs [hunk 8]
```diff
@@ -199,8 +199,8 @@ public PageValidationHelper(INcneUserDbService ncneUserDbService)
                 {
                     if (assignedUser != username)
                     {
-                        validationErrorMessages.Add(
-                            "Only users assigned to the V1 role are allowed to complete the workflow.");
+                        var msg = isWithdrawal ? "Only users assigned to the Verifier V1 role are allowed to complete the workflow." : "Only users assigned to the V1 role are allowed to complete the workflow.";
```
Comment (uhakdt) [inline src/NCNEPortal/Helpers/PageValidationHelper.cs:202]: Not sure about the messaging here: 
"to the Verifier V1 role are"
"to the V1 role are"
This second one - does it not have a more specific type or is it "Review and Assess"? Maybe Im not understanding it correctly though.

Comment (SamirHassoun) [reply src/NCNEPortal/Helpers/PageValidationHelper.cs:202]: V1 was a step in the old Withdrawal workflow, which does not exists in the new withdrawal.

Verify V1 is a user not a step; and hence a user assigned 'as' the Verify V1 is allowed to Complete the workflow

File: src/NCNEPortal/Data/TaskStageType.json (hunk 7 above)
Comment (SamirHassoun) [reply src/NCNEPortal/Data/TaskStageType.json:146]: Fixed, thanks

**Review by uhakdt — APPROVED** [review]

**Pull request merged**